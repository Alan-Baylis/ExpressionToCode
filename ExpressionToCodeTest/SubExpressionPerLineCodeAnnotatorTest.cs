using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ExpressionToCodeLib;
using Xunit;

namespace ExpressionToCodeTest
{
    public class SubExpressionPerLineCodeAnnotatorTest
    {
        static readonly ExpressionToCodeConfiguration config = ExpressionToCodeConfiguration.DefaultAssertionConfiguration.WithAnnotator(CodeAnnotators.SubExpressionPerLineCodeAnnotator);
        static readonly IAnnotatedToCode annotator = config.GetAnnotatedToCode();

        [Fact]
        public void AplusBapproved()
        {
            var a = 2;
            var b = 5;
            ApprovalTest.Verify(annotator.AnnotatedToCode(() => a + b));
        }

        [Fact]
        public void NestedArrayAccess()
        {
            var a = 2;
            var b = 5;
            var nums = Enumerable.Range(10, 10).ToArray();
            ApprovalTest.Verify(annotator.AnnotatedToCode(() => nums[a + b] < 7));
        }

        [Fact]
        public void NestedArrayAccessWithOuterAnd()
        {
            var a = 2;
            var b = 5;
            var nums = Enumerable.Range(10, 10).ToArray();
            ApprovalTest.Verify(annotator.AnnotatedToCode(() => a < b && nums[a + b] < 7 && b < 10));
        }

        [Fact]
        public void Binary_expressions_with_nesting()
        {
            var a = 2;
            var b = 5;
            var c = 3.45;
            var nums = Enumerable.Range(10, 10).ToArray();
            ApprovalTest.Verify(annotator.AnnotatedToCode(() => a < b && (c > -a || c > b) && b < 10));
        }

        [Fact]
        public void MethodCallsAndArrayLiterals()
        {
            var a = 2;
            var b = 5;
            ApprovalTest.Verify(annotator.AnnotatedToCode(() => Math.Max(a, b) > new[] { 3, 8, 13, 4 }.Average()));
        }

        [Fact]
        public void MessyStructureElidesNeatly() {
            var hmm = "1234567890";
            ApprovalTest.Verify(annotator.AnnotatedToCode(
                () => hmm[1] == hmm[2] || hmm[4] == hmm[int.Parse(hmm[8].ToString())] || false
                ));
        }
    }
}
