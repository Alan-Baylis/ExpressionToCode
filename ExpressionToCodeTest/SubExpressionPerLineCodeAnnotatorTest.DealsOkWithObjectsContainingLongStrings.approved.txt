assertion failed

new { A_long_string = string.Join("##", Enumerable.Range(0, 100)) + "suffix", A_short_string = "short", A_long_enumerable = Enumerable.Range(0, 1000) }
   →   "0##1##2##3##4##5##6##7##8##9##10##11##12##13##14##15##16##17##18##19##20##21##22##23##24##25##26##27##28##29##30##31##32##33##34##35##36##37##38##39# ... (caused assertion failure)

string.Join("##", Enumerable.Range(0, 100)) + "suffix"
     →   "0##1##2##3##4##5##6##7##8##9##10##11##12##13##14##15##16##17##18##19##20##21##22##23##24##25##26##27##28##29##30##31##32##33##34##35##36##37##38##39# ...

string.Join("##", Enumerable.Range(0, 100))
     →   "0##1##2##3##4##5##6##7##8##9##10##11##12##13##14##15##16##17##18##19##20##21##22##23##24##25##26##27##28##29##30##31##32##33##34##35##36##37##38##39# ...

Enumerable.Range(0, 100)
     →   { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, ... }

Enumerable.Range(0, 1000)
     →   { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, ... }

